text/0026-sassafras-consensus.md

```diff
@@ -0,0 +1,988 @@
+
+```rust
+    fn ring_vrf_verify(
+        signature: RingVrfSignature,
+        verifier: RingVerifier
+    ) -> Boolean;
+```
+
+In this document, the types `BandersnatchSecretKey`, `BandersnatchPublicKey`,
+`RingSignature`, `RingProof`, `RingProver` and `RingVerifier` are intentionally
+left undefined as are not relevant. Their definitions can be found in the
+`bandersnatch_vrfs` reference implementation.
+
+
+## 6. Sassafras Protocol
+
+### 6.1. Epoch's First Block
+
+The first block produced for an epoch `N` is required to include the descriptor
+for the next epoch `N+1`.
+
+The descriptor of next epoch is the `NextEpochDescriptor`.
+   
+```rust
+    AuthorityId ::= BandersnatchPublicKey;
+
+    Randomness ::= OCTET_STRING(SIZE(32));
+
+    NextEpochDescriptor ::= SEQUENCE {
+        authorities: SEQUENCE_OF AuthorityId,
+        randomness: Randomness,
+        configuration: ProtocolConfiguration OPTIONAL
+    }
+```
+
+- `authorities`: list of authorities for the next epoch.
+- `randomness`: randomness value associated to the next epoch.
+- `configuration`: next epoch optional protocol configuration.
+
+The `NextEpochDescriptor` must be `SCALE` encoded and embedded in the block
+header digest data.
+
+The identifier for the digest element is `BYTES("SASS")`.
+
+**Security Consideration**: Instances of `NextEpochDescriptor` are generated through
+on-chain code whenever a block is identified as the first of an epoch.
+Consequently, every node executing the block has the capability to verify if
+the descriptor generated during block execution matches the one produced by the
+block author, which is stored in the digest data.
+
+#### 6.1.1. Epoch Randomness
+
+Each epoch has an associated randomness value defined by the
+`NextEpochDescriptor` `Randomness` element.
+
+The first block of epoch `N` contains the randomness associated to the next epoch `N+1`.
+
+Randomness for epoch `N+1` is computed using the VRF output values which are associated
+to each block of epoch `N-1`.
+
+Assuming block `Bi` is produced during epoch `N-1`. Then the randomness for epoch `N+1`
+is incrementally generated as `RandomnessAccumulator = Blake2(Bi.VrfOut)'`.
+
+The first block produced for epoch `N` will propose as epoch `N+1` randomness the value
+of `RandomnessAccumulator` as found after the production of the last block associated
+to epoch `N-1`.
+
+#### 6.1.2. Protocol Configuration
+
+The `ProtocolConfiguration` primarily influences certain checks carried out
+during ticket validation. It is defined as follows:
+
+```rust
+    ProtocolConfiguration ::= SEQUENCE {
+        attempts_number: U32,
+        redundancy_factor: U32
+    }
+```
+
+- `attempts_number`: number of tickets that can be submitted by each next-epoch authority.
+- `redundancy_factor`: factor that enhances the likelihood of a candidate ticket
+  being deemed valid according to some protocol specific checks.
```

Comment:
I think this section should describe the consequences of the parameters. 
```
- `attempts_number` influences the anonymity of block producers.  All tickets have a public attempt number less than `attempts_number`.  If two tickets share an attempt number then adversaries know they must be by different block producers, which reduces anonymity late in the epoch.  We do not mind `max_attempts < epoch_length` though because this loss already becomes small when `attempts_number = 64` or `128`.  And larger requires more computation.
- `redundancy_factor` controls the expected number of extra tickets produced beyond `epoch_length`.  
```
I forget exactly what `redundancy_factor` really does right now, but we want it to be meaningful, not just 1 or 0.  An option goes like:

We need a probability `p` of a particular attempt yielding a valid ticket, so that roughly `n p - redundancy_factor * sigma > epoch_length` where `n = num_validators * attempts_number` and `sigma = n p (1-p)` is the standard deviation under the normal approximation.  Approximating `sigma` by `sigma' = n p' (1 - p')` with `p' = epoch_length / n` then `p = (epoch_length + redundancy_factor * sigma) / n`.  Aka there being aura-like blocks in the middle of the epoch is almost a `redundancy_factor` sigma event.